packages/flutter/lib/src/material/input.dart

```diff
@@ -95,49 +128,124 @@ class InputState extends ScrollableState<Input> {
+        themeData.text.caption.copyWith(color: focused ? focusHighlightColor : themeData.hintColor);
+
+      double topPaddingIncrement = themeData.text.caption.fontSize + (config.isDense ? 4.0 : 8.0);
+      double top = topPadding;
+      if (hasInlineLabel)
+        top += topPaddingIncrement + textStyle.fontSize - labelStyle.fontSize;
+
+      stackChildren.add(new AnimatedPositioned(
+        left: 0.0,
+        top: top,
+        duration: _kTransitionDuration,
+        curve: _kTransitionCurve,
+        child: new Text(config.labelText, style: labelStyle)
+      ));
+
+      topPadding += topPaddingIncrement;
     }
 
-    Color focusHighlightColor = themeData.accentColor;
-    Color cursorColor = themeData.accentColor;
-    if (themeData.primarySwatch != null) {
-      cursorColor = themeData.primarySwatch[200];
-      focusHighlightColor = focused ? themeData.primarySwatch[400] : themeData.hintColor;
+    if (config.hintText != null && _value.isEmpty && !hasInlineLabel) {
+      TextStyle hintStyle = themeData.text.subhead.copyWith(color: themeData.hintColor);
+      stackChildren.add(new Positioned(
+        left: 0.0,
+        top: topPadding + textStyle.fontSize - hintStyle.fontSize,
+        child: new Text(config.hintText, style: hintStyle)
+      ));
+    }
+
+    Color cursorColor = themeData.primarySwatch == null ?
+      themeData.accentColor :
+      themeData.primarySwatch[200];
+
+    EdgeDims margin = new EdgeDims.only(bottom: config.isDense ? 4.0 : 8.0);
+    EdgeDims padding = new EdgeDims.only(top: topPadding, bottom: 8.0);
+    Color borderColor = focusHighlightColor;
+    double borderWidth = focused ? 2.0 : 1.0;
+
+    if (config.errorText != null) {
+      borderColor = Colors.red[700];
+      borderWidth = 2.0;
+      if (!config.isDense) {
+        margin = const EdgeDims.only(bottom: 15.0);
+        padding = new EdgeDims.only(top: topPadding, bottom: 1.0);
+      }
     }
 
-    textChildren.add(new RawEditableLine(
-      value: _editableString,
-      focused: focused,
-      style: textStyle,
-      hideText: config.hideText,
-      cursorColor: cursorColor,
-      onContentSizeChanged: _handleContentSizeChanged,
-      scrollOffset: scrollOffsetVector
+    stackChildren.add(new AnimatedContainer(
+      margin: margin,
+      padding: padding,
+      duration: _kTransitionDuration,
+      curve: _kTransitionCurve,
+      decoration: new BoxDecoration(
+        border: new Border(
+          bottom: new BorderSide(
+            color: borderColor,
+            width: borderWidth
+          )
+        )
+      ),
+      child: new RawEditableLine(
+        value: _editableString,
+        focused: focused,
+        style: textStyle,
+        hideText: config.hideText,
+        cursorColor: cursorColor
+      )
     ));
 
+    if (config.errorText != null && !config.isDense) {
+      TextStyle errorStyle = themeData.text.caption.copyWith(color: Colors.red[700]);
```

Comment:
We could put that in ThemeData.  I'm not sure if it shows up elsewhere.  The spec doesn't say, but using digital color meter told me it was ren[700].